lib/spack/spack/build_environment.py

```diff
@@ -317,6 +318,17 @@ def _add_werror_handling(keep_werror, env):
     env.set("SPACK_COMPILER_FLAGS_REPLACE", " ".join(["|".join(item) for item in replace_flags]))
 
 
+def validate_compiler_for_pkg(pkg):
+    """Ensures selected compiler is actually capable of compiling
+    Fail early and provide an convenient warning in case it is not
+    """
+    compiler = pkg.compiler
+    for lang in ("cc", "cxx", "fc", "f77"):
+        comp = getattr(lang, compiler)
+        if comp:
+            compiler.try_compiler(lang)
```

Comment:
I think this might have some conceptual overlap with `asp.c_compiler_runs -> compiler.Compiler.compiler_verbose_output`

Although `c_compiler_runs` generally is only run on `linux` - when `using_libc_compatibility() -> True`. I think it would be reasonable to "promote" that and have it always run.

Reply:
I agree, I think the logic in `asp.CompilerParser` could use some integration here and IMO should completely replace my changes from this PR in `build_environment.py` as this is what I was trying to achieve with those changes and this is objectively better.

Will combine `compiler.Compiler._compile_dummy_c_source` with `try_compiler` (`_compile_dummy_c_source` should probably merge with `try_compiler`) and will update the logic in `asp.CompilerParser` accordingly